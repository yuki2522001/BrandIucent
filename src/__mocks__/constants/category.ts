import { Categories } from "@common-types/category";
import url from "@assets/images/icons/beverages.png";

const CATEGORY: Categories = {
  id: "1234",
  name: "Beverages",
  images: {
    src: url,
    alt: "This is icon beverages",
  },
};

const CATEGORY_MOCKING_LIST: Categories[] = [
  {
    id: "1234",
    name: "Beverages",
    images: {
      src: "@assets/images/icons/beverages.png",
      alt: "This is icon beverages",
    },
  },
  {
    id: "12345",
    name: "Dairy",
    images: {
      src: "@assets/images/icons/dairy.png",
      alt: "This is icon dairy",
    },
  },
  {
    id: "12345",
    name: "Meat",
    images: {
      src: "@assets/images/icons/meat.png",
      alt: "This is icon meat",
    },
  },
  {
    id: "12345",
    name: "Spices & Herbs",
    images: {
      src: "@assets/images/icons/spices-herbs.png",
      alt: "This is icon spices-herbs",
    },
  },
  {
    id: "12345",
    name: "Stationery",
    images: {
      src: "@assets/images/icons/stationery.png",
      alt: "This is icon stationery",
    },
  },
  {
    id: "12345",
    name: "Costmetic",
    images: {
      src: "@assets/images/icons/costmetic.png",
      alt: "This is icon costmetic",
    },
  },
];

export { CATEGORY, CATEGORY_MOCKING_LIST };
