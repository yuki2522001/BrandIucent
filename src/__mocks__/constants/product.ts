import { Product } from "@common-types/product";
import iceCreamSundae from "@assets/images/ice-cream-sundae.png";
import beefSteak from "@assets/images/beef-steak.png";

const PRODUCT_MOCKING: Product = {
  productId: "16466545464646547",
  categoryId: "1651454657522",
  productName: "Ice Cream Sundae",
  productDescription:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
  productQuantity: 5,
  originalPrice: {
    value: 20,
    currency: "$",
  },
  discountPrice: {
    value: 20,
    currency: "$",
  },
  offerQuantity: "3kg",
  offer: false,
  bestSelling: true,
  popular: false,
  images: {
    src: iceCreamSundae,
    alt: "This is ice-cream-sundae images",
  },
};
const PRODUCT_MOCKING_LIST: Product[] = [
  {
    productId: "16466545464646547",
    categoryId: "1651454657522",
    productName: "Ice Cream Sundae",
    productDescription:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
    productQuantity: 5,
    originalPrice: {
      value: 20,
      currency: "$",
    },
    discountPrice: {
      value: 20,
      currency: "$",
    },
    offerQuantity: "3kg",
    offer: false,
    bestSelling: true,
    popular: false,
    images: {
      src: iceCreamSundae,
      alt: "This is ice-cream-sundae images",
    },
  }
];

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
