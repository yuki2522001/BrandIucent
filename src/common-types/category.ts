export interface Categories {
  id: string;
  name: string;
  images: {
    src: string;
    alt: string;
  };
}