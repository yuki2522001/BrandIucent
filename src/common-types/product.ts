export interface Product {
  productId: string;
  categoryId: string;
  productName: string;
  productDescription: string ;
  productQuantity: number;
  originalPrice: {
    value: number;
    currency: string;
  };
  discountPrice: {
    value: number;
    currency: string;
  };
  offerQuantity: string;
  offer: boolean;
  bestSelling: boolean;
  popular: boolean;
  images: {
    src: string;
    alt: string;
  }
}

export interface ProductProps {
  product?: Product;
}