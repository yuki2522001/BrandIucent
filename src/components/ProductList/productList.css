/* style for product list */
.productList__normal {
  width: 384px;
  margin: 100px 70px 188px 0;
  background: var(--light);
  box-shadow: 0px 10px 40px rgba(0, 0, 0, 0.1);
}

.productList__medium {
  width: 368px;
  border-radius: 20px;
  margin: 50px 85px 164px 0;
  background: var(--light);
}

.productList__medium--box {
  width: 368px;
  background: var(--light);
  border-radius: 10px;
  margin: 0 80px 159px 0;
  box-shadow: 0px 16px 70px var(--box-first);
}

/* style for card in product list */
.cart__link {
  text-decoration: none;
  color: var(--text-primary);
}

.card__discount {
  padding: 59px 0 60px 44px;
}

.card__quantity {
  padding: 57px 0 15px 40px;
}

.card__unit {
  margin: 9px 0 14px;
  color: var(--bg-secondary);
}

.card__image {
  width: 368px;
  height: 288px;
}

.card__item--offers,
.card__item--selling,
.card__item--popular {
  display: flex;
  flex-wrap: wrap;
}

.card__item--offers {
  margin: 33px 0 0 90px;
}

.card__item--selling,
.card__item--popular {
  padding-left: 100px;
}

.card__counter {
  display: flex;
  justify-content: space-between;
  margin: 0 35px 18px;
}

.cardOffersList__title {
  margin: 33px 0 0 130px
}

.cardPopular__title {
  margin: 205px 0 44px 103px
}

.cardProductSelling__title {
  padding: 84px 0 49px 125px;
}

.cardProductSellingList {
  background: var(--bg-light);
}