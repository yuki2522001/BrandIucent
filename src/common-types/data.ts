import { Categories } from "./category";
import { Product } from "./product";

export interface DataContextProps {
  products: Product[];
  setProducts: Function;
  searchValue: [] | undefined;
  setSearchValue: Function;
  categories: Categories[];
  setCategories: Function;
}