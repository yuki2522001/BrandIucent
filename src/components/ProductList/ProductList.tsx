import React from "react";
import { Link } from "react-router-dom";
import { Product } from "@common-types/product";
import Text from "@components/common/Text/Text";
import Price from "@components/common/Price/Price";
import Counter from "@components/common/Counter/Counter";
import Icon from "@components/common/Icon/Icon";
import "./productList.css";

interface CartProductProps {
  type: string;
  content: string;
  visibleQuantity?: boolean;
  visibleDiscountPrice?: boolean;
  visibleCounter?: boolean;
  productList?: Product[];
}

const CardProductList: React.FC<CartProductProps> = ({
  type,
  content,
  visibleQuantity,
  visibleDiscountPrice,
  visibleCounter,
  productList,
}) => {
  let className = "";
  switch (type) {
    case "normal":
      className = "productList__normal";
      break;
    case "medium":
      className = "productList__medium";
      break;
    case "medium-box":
      className = "productList__medium--box";
      break;
    default:
      break;
  }

  let cartInfo = "";
  switch (content) {
    case "discount":
      cartInfo = "card__discount";
      break;
    case "quantity":
      cartInfo = "card__quantity";
      break;
    default:
      break;
  }

  return (
    <>
      {productList?.map((product: Product) => (
        <div
          data-testid="product-list"
          key={product.productId}
          className={className}
        >
          <img
            className="card__image"
            src={product.images.src}
            alt={product.images.alt}
          />
          <div className={cartInfo}>
            <Link className="cart__link" to={`/products/${product?.productId}`}>
              <Text size="normal" text={product?.productName} />
            </Link>
            {visibleQuantity && (
              <p className="card__unit">{product?.offerQuantity}</p>
            )}
            <Price
              type="original"
              value={product?.originalPrice.value}
              currency={product?.originalPrice.currency}
            />
            {visibleDiscountPrice && (
              <Price
                type="discount"
                value={product?.discountPrice.value}
                currency={product?.discountPrice.currency}
              />
            )}
          </div>
          {visibleCounter && (
            <div className="card__counter">
              <Counter counter={product.productQuantity} />
              <div className="card__icon">
                <Icon iconName="cart" />
              </div>
            </div>
          )}
        </div>
      ))}
    </>
  );
};

export default CardProductList;
