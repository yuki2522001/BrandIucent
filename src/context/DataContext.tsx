import { createContext, useMemo, useState } from "react";
import { Categories } from "@common-types/category";
import { DataContextProps } from "@common-types/data";
import { Product } from "@common-types/product";

const DataContext = createContext<DataContextProps>({} as DataContextProps);
const DataProvider: React.FC<{ children: JSX.Element[] | JSX.Element }> = ({
  children,
}) => {
  const [searchValue, setSearchValue] = useState<[]>([]);
  const [categories, setCategories] = useState<Categories[]>([]);
  const [products, setProducts] = useState<Product[]>([]);

  const value = {
    products,
    setProducts,
    searchValue,
    setSearchValue,
    categories,
    setCategories,
  };

  return <DataContext.Provider value={value}>{children}</DataContext.Provider>;
};

export { DataProvider, DataContext };
