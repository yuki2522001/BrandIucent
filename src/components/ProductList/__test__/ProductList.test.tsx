import mockAxios from "@__mocks__/axios";
import "@testing-library/jest-dom";
import ProductList from "../ProductList";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { DataContextProps } from "@common-types/data";
import { DataContext } from "@context/DataContext";
import { PRODUCT_MOCKING_LIST } from "@__mocks__/constants/product";
import { PRODUCTS_URL } from "@constants/url";
import { getData } from "@helpers/fetchApi";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/category";

const contextProductMock: DataContextProps = {
  searchValue: [],
  setSearchValue: jest.fn(),
  categories: CATEGORY_MOCKING_LIST,
  setCategories: jest.fn(),
  products: PRODUCT_MOCKING_LIST,
  setProducts: jest.fn(),
};

describe("CardProductList component", () => {
  afterEach(() => {
    mockAxios.reset();
  });

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: PRODUCT_MOCKING_LIST });
    const result = await getData(PRODUCTS_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(PRODUCTS_URL);
    expect(result).toEqual(PRODUCT_MOCKING_LIST);
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductList type="normal" content="normal" />
        </Router>
      </DataContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
