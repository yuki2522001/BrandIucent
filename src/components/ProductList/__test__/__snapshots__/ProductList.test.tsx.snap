// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProductList component matches snapshot 1`] = `<DocumentFragment />`;
